src/client/QXmppAccountMigrationManager.cpp

```diff
@@ -0,0 +1,336 @@
+// SPDX-FileCopyrightText: 2023 Filipe Azevedo <pasnox@gmail.com>
+// SPDX-FileCopyrightText: 2024 Filipe Azevedo <pasnox@gmail.com>
+// SPDX-FileCopyrightText: 2023 Linus Jahn <lnj@kaidan.im>
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "QXmppAccountMigrationManager.h"
+
+#include "QXmppClient.h"
+#include "QXmppPromise.h"
+#include "QXmppTask.h"
+#include "QXmppUtils_p.h"
+
+#include <QDomElement>
+
+// Utilities for account data extensions
+
+struct XmlElementId
+{
+    QString tagName;
+    QString xmlns;
+
+    bool operator==(const XmlElementId &other) const
+    {
+        return tagName == other.tagName && xmlns == other.xmlns;
+    }
+};
+
+template<>
+struct std::hash<XmlElementId>
+{
+    std::size_t operator()(const XmlElementId &m) const noexcept
+    {
+        QtPrivate::QHashCombine hash;
+        std::size_t seed = 0;
+        seed = hash(seed, m.tagName);
+        seed = hash(seed, m.xmlns);
+        return seed;
+    }
+};
+
+static std::unordered_map<XmlElementId, QXmppAccountData::AnyParser> &accountDataParsers()
+{
+    thread_local static std::unordered_map<XmlElementId, QXmppAccountData::AnyParser> registry;
+    return registry;
+}
+
+static std::unordered_map<std::type_index, QXmppAccountData::AnySerializer> &accountDataSerializers()
+{
+    thread_local static std::unordered_map<std::type_index, QXmppAccountData::AnySerializer> registry;
+    return registry;
+}
+
+struct QXmppAccountDataPrivate : public QSharedData
+{
+    QString sourceBareJid;
+    QVector<std::any> extensions;
+};
+
+QXMPP_PRIVATE_DEFINE_RULE_OF_SIX(QXmppAccountData)
+
+QXmppAccountData::QXmppAccountData(const QString &sourceBareJid)
+    : d(new QXmppAccountDataPrivate())
+{
+    setSourceBareJid(sourceBareJid);
+}
+
+std::variant<QXmppAccountData, QXmppError> QXmppAccountData::fromDom(const QDomElement &el)
+{
+    if (el.tagName().compare(u"accountData", Qt::CaseInsensitive) != 0) {
+        return QXmppError { QSL65("Not an accountData document"), {} };
+    }
+
+    const auto &parsers = accountDataParsers();
+    const auto nodes = el.childNodes();
+    QXmppAccountData data;
+
+    data.setSourceBareJid(el.attribute(QSL65("sourceBareJid")));
+
+    for (int i = 0; i < nodes.count(); ++i) {
+        const auto node = nodes.at(i);
+
+        if (!node.isElement()) {
+            qWarning("%s: Unexpected non element node (%ls)", Q_FUNC_INFO, qUtf16Printable(node.nodeName()));
+            continue;
+        }
+
+        const auto iq = node.toElement();
+        const auto element = iq.firstChildElement();
+        const auto parser = parsers.find(XmlElementId { element.tagName(), element.namespaceURI() });
```

Comment:
This shouldn't be limited to IQs and actually I would like to avoid to have IQs in `account-data` at all.
Better create new types with just the relevant information, we don't want to store an IQ with from/to/type, etc.

For the roster you could do:
```
struct RosterData
{
    QList<QXmppRosterIq::Item> items;
}
```

and use the parse/serialization of the roster items within a `<roster xmlns="org.qxmpp.accountdata">` wrapper element.

Reply:
Will do